Documented how to scale up / down ingesters (#3859)
cortexproject/cortex: docs/guides/ingesters-scaling-up-and-down.md

---
title: "Ingesters scaling up and down"
linkTitle: "Ingesters scaling up and down"
weight: 10
slug: ingesters-scaling-up-and-down
---

This guide explains how to scale up and down ingesters.

_If you're looking how to run ingesters rolling updates, please refer to the [dedicated guide](./ingesters-rolling-updates.md)._

## Scaling up

> Adding more ingesters to a Cortex cluster is considered a safe operation. When a new ingester startups, it will register to the [hash ring](../architecture.md#the-hash-ring) and the distributors will reshard received series accordingly.

Reviewer: ```suggestion
Adding more ingesters to a Cortex cluster is considered a safe operation. When a new ingester starts, it will register to the [hash ring](../architecture.md#the-hash-ring) and the distributors will reshard received series accordingly.
```

> No special care is required to take when scaling up ingesters.

Reviewer 2: I don't think this is true when running without shard-by-all-labels.

Each time you add an ingester some series move to the new ingester.
Suppose you have replication-factor=3, then a query will consult 3 ingesters who _now_ own the series and take the first 2 answers.
If you have added 2 or more ingesters recently then they will only have very recent data.  Medium-age data is still being uploaded to back-end store by the ingesters who previously owned the series. There will be a gap in the query result.

Reviewer 2: https://github.com/cortexproject/cortex/pull/3893

## Scaling down

> A running ingester holds several hours of time series data in memory, before they're flushed to the long-term storage.  When an ingester shutdowns, because of a scale down operation, the in-memory data must not be discarded in order to avoid any data loss.

Reviewer: ```suggestion
A running ingester holds several hours of time series data in memory, before they're flushed to the long-term storage.  When an ingester shuts down, because of a scale down operation, the in-memory data must not be discarded in order to avoid any data loss.
```

The procedure to adopt when scaling down ingesters depends on your Cortex setup:

- [Blocks storage](#blocks-storage)
- [Chunks storage with WAL enabled](#chunks-storage-with-wal-enabled)
- [Chunks storage with WAL disabled](#chunks-storage-with-wal-disabled-hand-over)

### Blocks storage

> When Cortex is running the [blocks storage](../blocks-storage/_index.md), ingesters don't flush series to blocks at shutdown by default. However, Cortex ingesters expose an API endpoint [`/shutdown`](../api/_index.md#shutdown) that can be called to flush series to blocks and upload them to the long-term storage before the ingester terminates.

Reviewer: ```suggestion
When Cortex is running the [blocks storage](../blocks-storage/_index.md), ingesters don't flush series to blocks at shutdown by default. However, Cortex ingesters expose an API endpoint [`/shutdown`](../api/_index.md#shutdown) that can be called to flush series to blocks and upload blocks to the long-term storage before the ingester terminates.
```

The ingesters scale down is deemed an infrequent operation and no automation is currently provided. However, if you need to scale down ingesters, please be aware of the following:

- Configure queriers and rulers to always query the storage
  - `-querier.query-store-after=0s`
- Frequently scan the storage bucket
  - `-blocks-storage.bucket-store.sync-interval=5m`
  - `-compactor.cleanup-interval=5m`
- Lower bucket scanning cache TTLs
  - `-blocks-storage.bucket-store.metadata-cache.bucket-index-content-ttl=1m`
  - `-blocks-storage.bucket-store.metadata-cache.tenant-blocks-list-ttl=1m`
  - `-blocks-storage.bucket-store.metadata-cache.metafile-doesnt-exist-ttl=1m`
- Ingesters should be scaled down one by one:
  1. Call `/shutdown` endpoint on the ingester to shutdown

>   2. Wait until the API terminates or "finished flushing and shipping TSDB blocks" is logged

Reviewer: ```suggestion
  2. Wait until the HTTP call returns successfully or "finished flushing and shipping TSDB blocks" is logged
```

  3. Terminate the ingester process (the `/shutdown` will not do it)

>   4. Wait 2x the maximum between `-blocks-storage.bucket-store.sync-interval` and `-compactor.cleanup-interval`

Reviewer: ```suggestion
  4. Before proceeding to the next ingester, wait 2x the maximum between `-blocks-storage.bucket-store.sync-interval` and `-compactor.cleanup-interval`
```

### Chunks storage with WAL enabled

When Cortex is running the [chunks storage](../chunks-storage/_index.md) with WAL enabled, ingesters don't flush series chunks to storage at shutdown by default. However, Cortex ingesters expose an API endpoint [`/shutdown`](../api/_index.md#shutdown) that can be called to flush chunks to the long-term storage before the ingester terminates.

The procedure to scale down ingesters -- one by one -- should be:

1. Call `/shutdown` endpoint on the ingester to shutdown

> 2. Wait until the API terminates or "flushing of chunks complete" is logged

Reviewer: ```suggestion
2.  Wait until the HTTP call returns successfully or "flushing of chunks complete" is logged
```
